arroyo-worker/src/connectors/kinesis/source/mod.rs

```diff
@@ -0,0 +1,508 @@
+                    }
+                    _ => return Err(anyhow!(error)),
+                },
+            }
+        }
+    }
+
+    async fn handle_get_records(
+        &mut self,
+        shard_id: String,
+        get_records: GetRecordsOutput,
+        ctx: &mut Context<(), T>,
+    ) -> Result<Option<BoxedFuture<AsyncNamedResult<AsyncResult>>>, UserError> {
+        let last_sequence_number = get_records.records().and_then(|records| {
+            records
+                .last()
+                .map(|record| record.sequence_number().unwrap().to_owned())
+        });
+
+        let next_shard_iterator = self.process_records(get_records, ctx).await?;
+        let shard_state = self.shards.get_mut(&shard_id).unwrap();
+
+        if let Some(last_sequence_number) = last_sequence_number {
+            shard_state.offset = KinesisOffset::SequenceNumber(last_sequence_number.to_string());
+        }
+
+        match next_shard_iterator {
+            Some(shard_iterator_id) => Ok(Some(self.next_read_future(shard_id, shard_iterator_id))),
+            None => {
+                shard_state.closed = true;
+                Ok(None)
+            }
+        }
+    }
+    async fn handle_need_new_iterator(
+        &mut self,
+        shard_id: String,
+    ) -> Result<Option<BoxedFuture<AsyncNamedResult<AsyncResult>>>, UserError> {
+        let get_shard_iterator_call = self.get_shard_iterator_call();
+        let shard_state = self.shards.get_mut(&shard_id).unwrap();
+        Ok(Some(
+            shard_state.update_shard_iterator(get_shard_iterator_call),
+        ))
+    }
+
+    fn get_shard_iterator_call(&self) -> GetShardIterator {
+        self.kinesis_client.as_ref().unwrap().get_shard_iterator()
+    }
+
+    async fn run_int(&mut self, ctx: &mut Context<(), T>) -> Result<SourceFinishType, UserError> {
+        let config = load_from_env().await;
+        self.kinesis_client = Some(KinesisClient::new(&config));
+        let mut futures = FuturesUnordered::new();
+        let get_shard_iterator_call = self.get_shard_iterator_call();
+        for shard in self.shards.values_mut() {
+            if shard.closed {
+                continue;
+            }
+            let get_shard_iterator_call = shard.set_shard_iterator(get_shard_iterator_call.clone());
+            futures.push(shard.update_shard_iterator(get_shard_iterator_call));
+        }
+
+        loop {
+            select! {
+                result = futures.select_next_some() => {
+                    let shard_id = result.name;
+                    match self.handle_async_result_split(shard_id,
+                        result.result.map_err(|e| UserError::new("Fatal Kinesis error", e.to_string()))?, ctx).await? {
+                        Some(future) => {
+                            futures.push(future);
+                        },
+                        None => {}
+                    }
+                },
+                _ = sleep(Duration::from_secs(60)) => {
```

Comment:
The logic here is that if we don't do anything else for 60 seconds, we'll call sync_shards. Is it intended that this never runs in steady-state, or is this meant to run every 60 seconds?

(Also generally some more comments would be great to explain how our code is interacting with the semantics of kinesis)

Reply:
No, it's meant to be something that should trigger about every 60 seconds. Should probably be using a tokio Interval.

Reply:
fixed.